### ISSUE TITLE
Blockwise observer ?

### ISSUE THREAD

**Issue by mb590721** [eclipse/californium#810]
Hello,

I'm using a coap resource to expose some data through a GET method. Because of the CoAP lib in the device software requests without blockwise need to be under 1024 bytes but this resource can sometime respond with more than that.

To avoid errors devices always send the request using blockwise, but I'm in front of another problem now. I need to make some treatments at the server side when the requests have been correctly send and I used to do it through a custom message observer thanks to the `onAcknowledgement` method.

If I understand correctly the BlockWise process, the server send the response with the blocksize and blocknumber specified by the client, and this client make a new requests for each remaining block. This explain that we never go inside the MessageObserver function. So my question is the following :

- How to know when the response has been completely send and received by the device ? I didn't find any BlockWise Observer.

We currently use the 2.0.0-M1 version of Californium. I think this may have been solved in later version (issue #316 ) but I would like to ensure there is no other way than upgrading version.

I hope being clear enough while describing my problem.

Thank you for your time !

**Comment by boaks** [comment]
Though critical bugs (leaks caused by threading issues) in blockwise and observe/notify are fixed in the meantime, I strongly recommend to update to 2.0.0-M12. There you find the `onComplete`, which indicates on the server side, that the last block is sent. The server never know, if the device really received that block. It may be assumed, that, if no retry arrives, the response block received the device, but that's an assumption.

**Comment by mb590721** [comment]
Okay thank you for this quick reply, the `onComplete` should be enough in our use case. I will check theses critical bugs to see if it impact us.

Is there a migration guide available ? Or do you have advices on critical point which could have changed between these version ?

I will see if I can find enough advantages to persuade to work on the migration.

**Comment by boaks** [comment]
No, we don't have migration guides for Milestones. And M1 is really long ago.

We introduced a CoapEndpoint.Builder and some additional DTLS configuration parameter.
And there are plenty of details fixed around the blockwise and observe/notifies, e.g. #527 or handlings #454.
I hope, 90% of your changes will just be some typing, and for the rest , please ask. 
And also hopefully, you have some tests, which cover your way of using it. We really fixed a lot around the blockwise, but especially, if it's used on a own interpretation, this may result in changed behavior.

The M13 will then come with some more changes in DTLS, but we ares sill working on that.

**Comment by mb590721** [comment]
Hi,

We upgraded to the M12 version but now devices fail to retrieve data through blockwise transfert with the block2 option.

I tried with the M1 californium client and it works fine so I assumed there was no breaking change. Do you know if CoAP message format changed between M1 and M12 ? On devices it seems to failed while reading the "m" flag which indicate if there are more blocks to request from the server.

**Comment by boaks** [comment]
> Do you know if CoAP message format changed between M1 and M12 ?

The message format was not changed. the only significant change, I'm aware of, was #527.
Before that, californium used the ETAG option to select different versions of a blockwise resource.
But that wasn't RFC7252 compliant. Maybe you can check, if your clients make use of the ETAG option (unfortunately in the none compliant usage :-) ).

> On devices it seems to failed while reading the "m" flag which indicate if there are more blocks to request from the server.

I'm not aware of a change in the m flag of option block2.
If possible, provide a capture of some successful single blockwise block requests and responses and with unsuccessful ones.

**Comment by mb590721** [comment]
I have some news : 

First of all at the server side I set the response as "CON" in order to have an ACK on simple requests : 

<pre>Response resp = new Response(ResponseCode.CONTENT);
resp.setConfirmable(true);</pre>

When using the M1 version on blockwise transfert this setting was overriden to send the response with ACK with same messageId instead of CON, which seem to respect the RFC (see https://tools.ietf.org/html/rfc7959#section-3.1). The client would then send another CON request asking for the second block and we would have keep up like that.

<pre>
>>> Datagram sent OK [30 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 2 Byte -> 0x0501
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 59992
----------------------------
<<< Datagram received [1034 Bytes]
----------------------------
*Type: ACK (0x02)
*Token: 2 Byte -> 0x0501
*Code: 2.05 (0x45)
*MessageId: 59992
----------------------------
Je-29/11/2018-15:26:14 - COAM - GET /blockwise (tkpsk) ... 2.05 / OK [813ms]
>>> Datagram sent OK [30 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 2 Byte -> 0x0601
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 59993
----------------------------
<<< Datagram received [368 Bytes]
----------------------------
*Type: ACK (0x02)
*Token: 2 Byte -> 0x0601
*Code: 2.05 (0x45)
*MessageId: 59993
----------------------------
Je-29/11/2018-15:26:14 - COAM - GET /blockwise (tkpsk) ... 2.05 / OK [303ms]
</pre>

When using the M12 version on blockwise transfert this setting seems to not be overriden and the server respond with a CON which has another messageId. I didn't see anything which forbide the use of "CON" response on the RFC for blockwise but as there is nothing that shows it I don't know if it's ok to do it.

In our case this fails when device request a resource with the block2 option but there is no need for further requests because a single response is enough to contain the payload. We haven't find yet why, it may be because the server didn't send the block2 options back.

Do you know if it's ok to send back the CON response for blockwise ? Should I change it manually to only send CON response for requests which didn't ask for blockwise ?

<pre>
>>> Datagram sent OK [47 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfb
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 30840
----------------------------
<<< Datagram received [814 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfb
*Code: 2.05 (0x45)
*MessageId: 25697
----------------------------
>>> Datagram sent OK [4 Bytes]
----------------------------
*Type: ACK (0x02)
*Token: 1 Byte -> 0xfb
*Code: 0.00 (0x00) [EMPTY]
*MessageId: 25697
----------------------------
Error code : 0x140E
</pre>

**Comment by boaks** [comment]
Would it be possible, that you print the block2 option and the etag option also?

The seconds case seems to be a request, but then I would guess you receive a CON notify (before the response).

**Comment by mb590721** [comment]
> Would it be possible, that you print the block2 option and the etag option also?

The development team in charge of this part is on it.

> The seconds case seems to be a request, but then I would guess you receive a CON notify (before the response).

I'm sorry I didn't understand what your were saying, is it possible to explain ?

Do you have any thoughts about the CON response on blockwise ?

**Comment by boaks** [comment]
Request, MID 30840, token FB
```
>>> Datagram sent OK [47 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfb
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 30840
----------------------------
```

Response, MID 25697, token FB
```
<<< Datagram received [814 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfb
*Code: 2.05 (0x45)
*MessageId: 25697
----------------------------
```

So, this may either be a separate CON response, where the ACK for CON request 30840 seems to be not included in the log. or, it is a CON notify. Printing the observe option would help here.

So, please print
- block2 option
- etag option
- observe option

of requests and responses.

**Comment by mb590721** [comment]
Request for blockwise with a payload that need to be sent over multiple packets.

<b> Client request </b>
<pre>
>>> Datagram sent OK [47 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfb
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 25294
*Option #11 (Length=4) ->[64][61][74][61]
*Option #11 (Length=2) ->[74][68]
*Option #15 (Length=4) ->[74][73][3d][30]
*Option #23 (Length=1) ->[06]
*Option #*** (Length=***) ->*******************
*Timestamp: 1543506281
----------------------------
</pre>

<b>Server response</b>
<pre>
<<< Datagram received [1050 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfb
*Code: 2.05 (0x45)
*MessageId: 25733
*Option #23 (Length=1) ->[0e]
*Option #28 (Length=2) ->[06][0e]
*Option #*** (Length=***) ->*******************
*Payload (1024 Byte): "***************************"
*Timestamp: 1543506282
----------------------------
</pre>
<b>Client ACK CON response</b>
<pre>
>>> Datagram sent OK [4 Bytes]
----------------------------
*Type: ACK (0x02)
*Token: 1 Byte -> 0xfb
*Code: 0.00 (0x00) [EMPTY]
*MessageId: 25733
*Timestamp: 1543506282
----------------------------
</pre>

<b>Client request for next bloc</b>
<pre>
>>> Datagram sent OK [34 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 1 Byte -> 0xfc
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 25295
*Option #11 (Length=4) ->[64][61][74][61]
*Option #11 (Length=2) ->[74][68]
*Option #15 (Length=4) ->[74][73][3d][30]
*Option #23 (Length=1) ->[1e]
*Option #*** (Length=***) ->*******************
*Timestamp: 1543506282
----------------------------
</pre>
<b>Server respond with ACK for rest of the blockwise</b>
<pre>
<<< Datagram received [549 Bytes]
----------------------------
*Type: ACK (0x02)
*Token: 1 Byte -> 0xfc
*Code: 2.05 (0x45)
*MessageId: 25295
*Option #23 (Length=1) ->[16]
*Option #*** (Length=***) ->*******************
*Payload (526 Byte): "********************"
*Timestamp: 1543506282
----------------------------
</pre>

So everthing work fine in this case. 

But when we make the same request but with a payload size which can be send with the first response (no need to get more blocks) we see that the server <b>doesn't send any option</b>. We use custom options which are send in every exchange so it's not just the block2 option.

<b>Client request</b>
<pre>
>>> Datagram sent OK [30 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 2 Byte -> 0x0701
*Code: 0.01 (0x01) [REQ_GET]
*MessageId: 25306
*Option #11 (Length=4) ->[64][61][74][61]
*Option #11 (Length=2) ->[74][68]
*Option #23 (Length=1) ->[0e]
*Option #***** (Length=*****) ->*********************
*Timestamp: 1543510614
----------------------------
</pre>
<b>Server response</b>
<pre>
<<< Datagram received [633 Bytes]
----------------------------
*Type: CON (0x00)
*Token: 2 Byte -> 0x0701
*Code: 2.05 (0x45)
*MessageId: 25740
*Payload (626 Byte): "*********"
*Timestamp: 1543510614
----------------------------
</pre>
<b> Client ACK CON response of server </b>
<pre>
>>> Datagram sent OK [4 Bytes]
----------------------------
*Type: ACK (0x02)
*Token: 2 Byte -> 0x0701
*Code: 0.00 (0x00) [EMPTY]
*MessageId: 25740
*Timestamp: 1543510614
----------------------------
</pre>

And the client library failed after that because it tries to read the block2 option to see if there are more blocks to get, but there isn't any option.
<pre>
Error code : 0x140E
</pre>


I think there is 2 points to figure out :

- why server doesn't send any options on blockwise when only one response in necessary. (probably a bug)
- is it RFC compliant to respond to a blockwise request with CON response? It seems it's just the first response which is send as a CON, the ones after that are send as ACK.

What do you think ?

**Comment by mb590721** [comment]
Maybe we should open another issue to track it ? I don't think that this is related to this one.

**Comment by boaks** [comment]
> why server doesn't send any options on blockwise when only one response in necessary. (probably a bug)

I reread RFC7959, https://tools.ietf.org/pdf/rfc7959.pdf#18.
But it doesn't contain a direct definition for the case, where the request has a block 2 option, but the response doesn't require a block 2 option, because it just fits into a single response. There are just some parts, which could be interpreted :-).
Anyway, I will try to make it configurable, that the block2 is used in the response, even, if it fits into a single block. Otherwise, I guess, you will not be the only one, who complains ...

> is it RFC compliant to respond to a blockwise request with CON response? It seems it's just the first response which is send as a CON, the ones after that are send as ACK.
 
If  I use the SimpleFileServer, I get ACK.
If I introduce a `CoapExchange.accept()` before `response()`

```
exchange.accept();
exchange.respond(response);
```

I get in wireshark:
[accepted-block2.zip](https://github.com/eclipse/californium/files/2630313/accepted-block2.zip)
 
CON GET 
ACK EMPTY
CON 2.05
ACK EMPTY

So, I guess you use `accept()` and you don't print the ACK EMPTY. 
May be you can try to use wireshark for capture?

**Comment by mb590721** [comment]
line 556 of the BlockwiseLayer :

`else if (requiresBlockwise(exchange, response, requestBlock2))` 

<pre>
	private boolean requiresBlockwise(final Exchange exchange, final Response response, final BlockOption requestBlock2) {

		boolean blockwiseRequired = response.getPayloadSize() > maxMessageSize;
		if (requestBlock2 != null) {
			// client might have included early negotiation block2 option
			blockwiseRequired = blockwiseRequired || response.getPayloadSize() > requestBlock2.getSize();
		}
		if (blockwiseRequired) {
			LOGGER.debug("response body [{}/{}] requires blockwise transfer", response.getPayloadSize(),
					maxMessageSize);
		}
		return blockwiseRequired;
	}
</pre>

If the payload size is lower than the size asked by the client, we do not respond with blockwise. In this case no block2 option is inserted and that confuse the client which expects it to be present on the response to know if there is more blocks to fetch.

I thought of something like :

<pre>
private boolean requiresBlockwise(final Exchange exchange, final Response response, final BlockOption requestBlock2) {

	if(requestBlock2 != null){
		LOGGER.debug("client asked for early blockwise transfer");
	}else if(response.getPayloadSize() > maxMessageSize){
		LOGGER.debug("response body [{}/{}] requires blockwise transfer", response.getPayloadSize(),
				maxMessageSize);	
	}
	
	return blockwiseRequired;
}
</pre>

I may have miss something, it's only a suggestion.

**Comment by mb590721** [comment]
> why server doesn't send any options on blockwise when only one response in necessary. (probably a bug) 

I made some debug on my side and options are correctly send, it just the block2 option which isn't on this specific use case.


I will check your others comments tomorrow :-)

**Comment by boaks** [comment]
> I may have miss something, it's only a suggestion.

I already had a look on it. The issue will be,  that I only want this behavior, if it's explicitly configured. I expect, that clients usually do not fail, if they don't have a block2 option in short responses. So this setting will only be for "trouble cases". 

@sbernard31 
What do you think? Just add block 2 option, or make it configurable?

**Comment by boaks** [comment]
For a fast fix, you may try, line 1146:
``` 
private boolean requiresBlockwise(final Exchange exchange, final Response response, final BlockOption requestBlock2) {

boolean blockwiseRequired = requestBlock2 != null || response.getPayloadSize() > maxMessageSize;
```

Add the `requestBlock2 != null ||`

**Comment by mb590721** [comment]
> What do you think? Just add block 2 option, or make it configurable?

For either solution I could work on a PR if your are okay with that :-).

**Comment by boaks** [comment]
Sure, if you have a https://www.eclipse.org/legal/ECA.php

**Comment by mb590721** [comment]
> Sure, if you have a https://www.eclipse.org/legal/ECA.php

I will fulfill the form in the following days in order to work on it next week.

> According:
https://tools.ietf.org/pdf/rfc7959.pdf#12

Thanks for pointing this part out, I totally missed it. 

I agree that a fix should be made on the client library to handle this specific case, I will forward that to the team in charge of it to.

> By the way, you don't need to include a block2 option with sz 6 (1024) and num 0 in the initial request, the server will not chose a larger size anyway and it start a bockwise simply if the body is larger then the preferred size..

We will also check that point, if I remember well it was due to a limitation in the client library that it was done like that.
If I understand well the early negociation is only required when the client wants to get blocks with a size < 1024, that's right ?

**Comment by boaks** [comment]
> If I understand well the early negociation is only required when the client wants to get blocks with a size < 1024, that's right ?

Yes, your right, that's why "sz 6, block 0" could be left out from the initial request.

**Comment by boaks** [comment]
> I will fulfill the form in the following days in order to work on it next week.

You may require the permission of you employer (at least informational).

I would really welcome, if you can check ahead the fix above, if that would work for you.

line 1146:
```
private boolean requiresBlockwise(final Exchange exchange, final Response response, final BlockOption requestBlock2) {

boolean blockwiseRequired = requestBlock2 != null || response.getPayloadSize() > maxMessageSize;
```

Add the `requestBlock2 != null ||`

Not that this is only the first of a sequence of issues ...

**Comment by boaks** [comment]
Any progress? Could you verify, that "this fix is the fix" or "a step to the next issue"?

**Comment by mb590721** [comment]
Hi,

I began to work on it on my free time, I made the change and add a test case to ensure that it works. While adding the test I found out that when the blockwise layer retrieve the last block (when receiving a blockwise response so) it build the final response object without including the last block2 option, so when a checked the complete response my test failed because of it. I will use the same config field to add (or not) the block2 option in this case.

Do you want me to make a pull request for you to check it out even if I have not totally ended it?

**Comment by boaks** [comment]
> it build the final response object without including the last block2 option

Yes, that's intended, the "complete response" doesn't contain the block2 option.

If you just want a client, which checks, that the block 2 option is contained, set the  "PREFERRED_BLOCK_SIZE" to 0 on the client-side. If the payload on the server fits into the server's  
PREFERRED_BLOCK_SIZE, the client will receive a single block response with a block 2.

**Comment by boaks** [comment]
MID    : 49446
Token  : [b24ac49af84099a0]
Type   : ACK
Status : 2.05
Options: {"Content-Format":"application/octet-stream", "Size2":4}
RTT    : 17 ms
Payload: 4 Bytes

**Comment by boaks** [comment]
Uups! No, not the PREFERRED_BLOCK_SIZE, the MAX_RESOURCE_BODY_SIZE at the client-side must be 0 :-).

MID    : 43556
Token  : [d45c30fc5bc54f0d]
Type   : CON
Status : 2.05
Options: {"Content-Format":"application/octet-stream", "Block2":"(szx=6/1024, m=false, num=0)", "Size2":5}
RTT    : 17 ms
Payload: 5 Bytes

**Comment by mb590721** [comment]
Oh great tips, my tests passed with it! I just see that it disabled the transparent blockwise process :-).

I should be able to make a PR by the end of the week.

**Comment by boaks** [comment]
Great! I'm looking forward to it.
If you use the fix, is your issue then solved? Does blockwise then work for you "for now"?

**Comment by mb590721** [comment]
I didn't test it with the embedded lib but yes it should work without problem with this fix and allow us to migrate to the latest Californium version.

I'll keep you up to date.

**Comment by mb590721** [comment]
I didn't have the time to check with our embedded lib yet but in the meantime I made a PR to let you check what I've done so far.

**Comment by sbernard31** [comment]
Uuupps, I missed I was pinged on this thread (https://github.com/eclipse/californium/issues/810#issuecomment-442937936) ...

My feeling on this (which is pretty much the same as @boaks)

1. The spec specifies how an absence of block option should be handled (https://tools.ietf.org/html/rfc7959#section-2.2). So a CoAP implementation should handle it.  I would say there is no issue in Californium about not adding block2 option
2. I'm not against adding a configuration about that. (unfixable buggy devices is real life)
3. I agree that sending an early block2 negotiation with a 1024 value does not really make sense as 1024 is max value.
4. In your example https://github.com/eclipse/californium/issues/810#issuecomment-442905137, your client send a CON request and you directly answer with CON response. This sounds strange to me.
If you are able to answer quickly, I can not see the benefits of using separated response instead of piggyback. If there is a large delay between request and response, using separated response could make sense but in this case a ACK should be sent quickly (before the delayed CON response)

**Comment by boaks** [comment]
> your client send a CON request and you directly answer with CON response. This sounds strange to me.

I guess, the log just doesn't contain the ACK, which californium sends.
https://github.com/eclipse/californium/issues/810#issuecomment-442931461

My assumption:
Using a separate response enables the coap-server to ensure, that the response is send successful. But in this case it is unfortunately not working in that way, because the `accept` should  be used for the last block, not the first, but that is currently not implemented.
I guess, this is done to release "some buffer" and the switch to the next buffer with newly arrived data. But for now, focus on #821.

**Comment by sbernard31** [comment]
> Using a separate response enables the coap-server to ensure, that the response is send successful. 

Generally using piggyback means that if response is lost, client will send the request again (coap retransmission) so you should not lose the response. Except you are using a really unstable network (in that case request will timeout) or you plan to loose your coap state just after client receive the response like a reboot.

> I guess, this is done to release "some buffer" and the switch to the next buffer with newly arrived data.

I never thought about using it to release memory sooner. It could make sense.

**Comment by mb590721** [comment]
> My assumption:
Using a separate response enables the coap-server to ensure, that the response is send successful. But in this case it is unfortunately not working in that way, because the accept should be used for the last block, not the first, but that is currently not implemented.

Yes that was why I used a CON response. I make some specific treatments when I have the confirmation that the response was successfully received by the client.

It works well when the content fits into a single response but like you said when it comes to blockwise response it losts some sense.
On blockwise transfert the better approach that I saw is to use the `onComplete` method of the `MessageObserver` which is triggered when the last block is sent.

It would be a nice enhancement to add a callback in the `MessageObserver` on acknowledgement of the last block. Maybe I will work on it when I will have some free time.

By the way any idea when the M13 will be released ?
